File: packages/disk/src/archil.ts
import { type ApiClient, createApiClient, unwrap } from "./client.js";
import { Disk } from "./disk.js";
import { Disks } from "./disks.js";
import { Tokens } from "./tokens.js";
import { Workspace } from "./workspace.js";
import { deriveS3BaseUrl, resolveBaseUrl } from "./regions.js";
import type { ExecDiskResult } from "./types.js";

export interface ArchilOptions {
  /** API key. Falls back to ARCHIL_API_KEY env var if not provided. */
  apiKey?: string;
  /** Region. Falls back to ARCHIL_REGION env var if not provided. */
  region?: string;
  /** Override the control plane base URL (useful for testing). */
  baseUrl?: string;
  /**
   * Override the S3-compatible API base URL used by Disk#getObject /
   * putObject / deleteObject. Falls back to ARCHIL_S3_BASE_URL, then to the
   * control plane URL with its `control.` hostname prefix swapped for `s3.`.
   */
  s3BaseUrl?: string;
}

/**
 * Options that apply to a single mounted disk in an exec request.
 * Use this object form when you need to pin the mount to a subdirectory
 * of the disk, mount it read-only, mount it in conditional mode, or request
 * delegation checkouts; for the default case (mount the disk's root,
 * read-write with no checkout), pass a `Disk` or disk-id string instead.
 */
export interface ExecMountSpec {
  /** Disk to mount, by `Disk` instance or raw disk id string. */
  disk: Disk | string;
  /**
   * Subdirectory of the disk to expose at the mountpoint. Must be a
   * relative path with no `.` or `..` segments. When omitted, the disk's
   * root is exposed.
   */
  subdirectory?: string;
  /**
   * When true, mount the disk read-only inside the container. Writes
   * against the mount fail with EROFS. Defaults to false.
   */
  readOnly?: boolean;
  /**
   * When true, mount the disk in conditional mode, where mutating operations
   * are sent directly to the server without a delegation checkout. This
   * enables concurrent writes from multiple clients to the same disk.
   * Defaults to false.
   */
  conditional?: boolean;
  /**
   * Milliseconds to wait in the delegation queue for each requested checkout.
   * If set without `checkoutPaths`, the exposed mount root is acquired during
   * mount setup. Cannot be combined with `readOnly: true`.
   */
  queueMs?: number;
  /**
   * Paths relative to this disk's exposed mount root to check out before the
   * command starts. May be set without `queueMs`; those checkouts try
   * immediately without waiting in the delegation queue. Cannot be combined
   * with `readOnly: true`.
   */
  checkoutPaths?: string[];
}

/**
 * One disk to mount in an exec request. Either a `Disk`/disk-id string
 * (mounts the disk's root, read-write) or an `ExecMountSpec` object that
 * additionally selects a subdirectory of the disk and/or marks the mount as
 * read-only, conditional, or requests delegation checkouts. Used
 * by Archil#exec, where the map key is the relative path under /mnt/archil at
 * which to mount the disk.
 */
export type ExecMount = Disk | string | ExecMountSpec;

export interface ExecOptions {
  /**
   * Disks to mount, keyed by the relative path under `/mnt/archil` at which
   * to mount each one. At least one entry is required. Paths must be
   * non-empty, non-absolute, and contain no `.` or `..` segments.
   */
  disks: Record<string, ExecMount>;
  /** Shell command to run inside the container. */
  command: string;
}

/**
 * Read an environment variable in a way that is safe in browsers, where the
 * `process` global does not exist. Returns undefined when there is no process
 * environment, so the SDK can be bundled and run client-side as long as
 * credentials are passed explicitly to `new Archil({ ... })`.
 */
function envVar(name: string): string | undefined {
  return typeof process !== "undefined" && process.env ? process.env[name] : undefined;
}

function isExecMountSpec(m: ExecMount): m is ExecMountSpec {
  return typeof m === "object" && m !== null && "disk" in m;
}

function diskIdFromMount(m: Disk | string): string {
  return typeof m === "string" ? m : m.id;
}

export class Archil {
  readonly disks: Disks;
  readonly tokens: Tokens;
  /** @internal */
  private readonly _client: ApiClient;

  constructor(opts: ArchilOptions = {}) {
    const apiKey = opts.apiKey ?? envVar("ARCHIL_API_KEY");
    const region = opts.region ?? envVar("ARCHIL_REGION");

    if (!apiKey) {
      throw new Error("Missing API key: pass apiKey in options or set ARCHIL_API_KEY environment variable");
    }
    if (!region) {
      throw new Error("Missing region: pass region in options or set ARCHIL_REGION environment variable");
    }

    // Resolve the control-plane URL the same way the client does (explicit
    // override, else region lookup) so the S3 endpoint can be derived from it
    // even on the common region-based path.
    const controlBaseUrl = opts.baseUrl ?? resolveBaseUrl(region);

    const client = createApiClient({
      apiKey,
      region,
      baseUrl: controlBaseUrl,
    });

    // Derive the S3 endpoint from the control plane URL when not given
    // explicitly (swap the leading `control.` hostname segment for `s3.`).
    const s3BaseUrl =
      opts.s3BaseUrl ?? envVar("ARCHIL_S3_BASE_URL") ?? deriveS3BaseUrl(controlBaseUrl);

    this._client = client;
    this.disks = new Disks(client, region, s3BaseUrl);
    this.tokens = new Tokens(client);
  }

  /**
   * Run a command in a container with multiple disks mounted simultaneously,
   * each at its own relative path under `/mnt/archil`. Blocks until the
   * command completes and returns its stdout, stderr, exit code, and timing.
   */
  async exec(opts: ExecOptions): Promise<ExecDiskResult> {
    type ExecDiskWire = {
      disk: string;
      subdirectory?: string;
      readOnly: boolean;
      conditional: boolean;
      queueMs?: number;
      checkoutPaths?: string[];
    };
    const disks: Record<string, string | ExecDiskWire> = {};
    for (const [relPath, mount] of Object.entries(opts.disks)) {
      if (isExecMountSpec(mount)) {
        const entry: ExecDiskWire = {
          disk: diskIdFromMount(mount.disk),
          readOnly: mount.readOnly ?? false,
          conditional: mount.conditional ?? false,
        };
        if (mount.subdirectory !== undefined) entry.subdirectory = mount.subdirectory;
        if (mount.queueMs !== undefined) {
          entry.queueMs = mount.queueMs;
        }
        if (mount.checkoutPaths !== undefined) {
          entry.checkoutPaths = mount.checkoutPaths;
        }
        disks[relPath] = entry;
      } else {
        disks[relPath] = diskIdFromMount(mount);
      }
    }
    return unwrap<ExecDiskResult>(
      this._client.POST("/api/exec", {
        body: { disks, command: opts.command },
      }),
    );
  }

  /**
   * Build an agent filesystem toolset spanning several disks at once. `mounts`
   * maps a relative path to a disk (or `ExecMountSpec`), exactly like
   * {@link exec}; each disk appears under `/mnt/archil/<path>`. Hand the result
   * to a framework adapter (`agentTools(workspace)`) to get tools that route
   * file operations by path and fan `grep`/`list_files` out across the disks.
   */
  workspace(mounts: Record<string, ExecMount>): Workspace {
    return new Workspace(this, mounts);
  }
}

File: packages/python/archil/_archil.py
from __future__ import annotations

import os
from dataclasses import dataclass
from typing import Optional, Union

from ._disks import _Disks
from ._http import _Transport
from ._models import ExecResult
from ._regions import derive_s3_base_url, resolve_base_url
from ._tokens import _Tokens

# Imported at runtime (not under TYPE_CHECKING) so the synchronicity stub
# generator can resolve workspace()'s return type. _workspace imports _archil
# only lazily (inside methods), so this does not create an import cycle.
from ._workspace import _Workspace


@dataclass
class ExecMountSpec:
    """One disk to mount in an exec request, with optional subdirectory pinning,
    a read-only flag, and a conditional-mode flag. For the default case (mount
    the disk's root, read-write), pass a ``Disk`` or disk-id string directly
    instead.

    ``conditional`` mounts the disk in conditional mode, where mutating
    operations are sent directly to the server without a delegation checkout.
    This enables concurrent writes from multiple clients to the same disk."""

    disk: object  # a Disk (wrapped or impl) or a disk-id string
    subdirectory: Optional[str] = None
    read_only: bool = False
    conditional: bool = False


# A Disk instance, a disk-id string, or an ExecMountSpec.
ExecMount = Union[object, str, ExecMountSpec]


def _disk_id(mount: object) -> str:
    if isinstance(mount, str):
        return mount
    # Both the wrapped Disk and the impl _Disk expose `.id`.
    return mount.id  # type: ignore[attr-defined]


class _Archil:
    """Top-level Archil client. Holds the account-level ``disks`` and ``tokens``
    collections and the cross-disk ``exec``. Construct directly for multi-account
    or multi-region scripts; otherwise use the module-level helpers.

    Every method is available both synchronously and asynchronously: call it
    directly to block, or use the ``.aio`` attribute for a coroutine
    (e.g. ``await archil.exec.aio(...)``). Also usable as an (async) context
    manager: ``with Archil(...) as a:`` / ``async with Archil(...) as a:``."""

    def __init__(
        self,
        *,
        api_key: Optional[str] = None,
        region: Optional[str] = None,
        base_url: Optional[str] = None,
        s3_base_url: Optional[str] = None,
        timeout: Optional[float] = 30.0,
        _http_transport=None,
    ) -> None:
        api_key = api_key or os.environ.get("ARCHIL_API_KEY")
        region = region or os.environ.get("ARCHIL_REGION")
        if not api_key:
            raise ValueError(
                "Missing API key: pass api_key or set the ARCHIL_API_KEY environment variable"
            )
        if not region:
            raise ValueError(
                "Missing region: pass region or set the ARCHIL_REGION environment variable"
            )

        # Resolve the control-plane URL the same way the transport would (explicit
        # override, else region lookup) so the S3 endpoint can be derived from it
        # even on the common region-based path.
        control_base_url = base_url or resolve_base_url(region)
        s3 = s3_base_url or os.environ.get("ARCHIL_S3_BASE_URL") or derive_s3_base_url(control_base_url)

        self._transport = _Transport(
            control_base_url, api_key, s3, transport=_http_transport, timeout=timeout
        )
        self._disks = _Disks(self._transport, region)
        self._tokens = _Tokens(self._transport)

    @property
    def disks(self) -> "_Disks":
        return self._disks

    @property
    def tokens(self) -> "_Tokens":
        return self._tokens

    async def exec(self, *, disks: dict[str, ExecMount], command: str) -> ExecResult:
        """Run a command in a container with multiple disks mounted
        simultaneously, each at its own relative path under ``/mnt/archil``.
        Blocks until the command completes and returns its stdout, stderr, exit
        code, and timing."""
        payload: dict[str, object] = {}
        for rel_path, mount in disks.items():
            if isinstance(mount, ExecMountSpec):
                entry: dict[str, object] = {
                    "disk": _disk_id(mount.disk),
                    "readOnly": mount.read_only,
                    "conditional": mount.conditional,
                }
                if mount.subdirectory is not None:
                    entry["subdirectory"] = mount.subdirectory
                payload[rel_path] = entry
            else:
                payload[rel_path] = _disk_id(mount)
        data = await self._transport.request_json(
            "POST",
            "/api/exec",
            json={"disks": payload, "command": command},
            timeout=None,
        )
        return ExecResult.from_json(data)

    def workspace(self, mounts: dict[str, ExecMount]) -> "_Workspace":
        """Build a :class:`Workspace`: a filesystem spanning several disks at once.

        ``mounts`` maps a relative path to a disk (or ``ExecMountSpec``), exactly
        like :meth:`exec`; each disk appears as a top-level directory. The
        workspace is a full filesystem — ``read_file`` / ``write_file`` /
        ``list_files`` / ``grep`` / ``exec`` route to the right disk by path, and
        ``grep`` / ``list_files`` fan out across all of them — and
        ``agent_tools()`` builds drop-in agent tools just like :meth:`Disk`::

            ws = archil.workspace({"data": disk_a, "cache": disk_b})
            agent = Agent(tools=ws.agent_tools().for_langchain())
        """
        return _Workspace(self, mounts)

    async def close(self) -> None:
        """Close the underlying HTTP connections. Optional — also usable as a
        context manager (``with Archil(...) as archil:``)."""
        await self._transport.aclose()

    async def __aenter__(self) -> "_Archil":
        return self

    async def __aexit__(self, exc_type, exc, tb) -> None:
        await self._transport.aclose()
